### DISCUSSION TITLE
Restrict interface availability depending on parameter

### DISCUSSION THREAD

**Discussion by be5invis** [dotnet/csharplang#551] [Language Ideas]
Objective: Let `List<T>` implement `IEquatable<List<T>>` **if and only if** `T` implements `IEquatable<T>`.
Like in Haskell:
```haskell
instance (Eq a) => Eq [a] where ...
```

**Comment by casperOne** [comment]
I definitely don't want this.  I'd rather have it in another type extending `List<T>` which has equality semantics.

The question of ordering and how it relates to equality isn't addressed here either.

**Comment by HaloFour** [comment]
The CLR has no facility to allow conditional interfaces at all, let alone driven by a generic type argument.  How would you propose such a feature be implemented in CLR metadata?

**Comment by alrz** [comment]
Related: https://github.com/dotnet/csharplang/issues/164
```swift
extension class ListExtensions<T> : List<T>, IEquatable<List<T>> where T : IEquatable<T> { ... }

// alternative syntax, just for the sake of readability
implement<T> IEquatable<List<T>> for List<T> where T : IEquatable<T> { .. }
```
However, the proposal does not cover implementing regular interfaces (because it most likely requires clr support, along with "default interface methods", maybe), but I think that's definitely on the table.
